src/apparmor.rs

```diff
@@ -0,0 +1,66 @@
+use std::ffi::{c_char, c_int, CStr, CString};
+use std::{fs, io, mem};
+
+use crate::cutils::cerr;
+
+/// Set the profile for the next exec call if AppArmor is enabled
+pub fn set_profile_for_next_exec(profile_name: &str) -> io::Result<()> {
+    if apparmor_is_enabled()? {
+        apparmor_prepare_exec(profile_name)
+    } else {
+        // if the sysadmin doesn't have apparmor enabled, fail softly
+        Ok(())
+    }
+}
+
+fn apparmor_is_enabled() -> io::Result<bool> {
+    match fs::read_to_string("/sys/module/apparmor/parameters/enabled") {
+        Ok(enabled) => Ok(enabled.starts_with("Y")),
+        Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(false),
+        Err(e) => Err(e),
+    }
+}
+
+/// Switch the apparmor profile to the given profile on the next exec call
+fn apparmor_prepare_exec(new_profile: &str) -> io::Result<()> {
+    // SAFETY: Always safe to call
+    unsafe { libc::dlerror() }; // Clear any existing error
+
+    // SAFETY: Loading a known safe dylib. LD_LIBRARY_PATH is ignored because we are setuid.
+    let handle = unsafe { libc::dlopen(c"libapparmor.so.1".as_ptr(), libc::RTLD_NOW) };
+    if handle.is_null() {
+        // SAFETY: In case of an error, dlerror returns a valid C string.
+        return Err(io::Error::new(io::ErrorKind::NotFound, unsafe {
+            CStr::from_ptr(libc::dlerror())
+                .to_string_lossy()
+                .into_owned()
+        }));
+    }
+
+    // SAFETY: dlsym will either return a function pointer of the right signature or NULL.
+    // Option<fn()> is guaranteed to represent a NULL value as None and any other value as Some.
```

Comment:
Making a pedantic point here:

```rust
transmute::<_, Option<fn()>>([0u8; size_of::<fn()>()])
```
is sound per the Rust reference and will produce `None`, but as far as my C bible says, `NULL` is not guaranteed to be [0u8; size_of::<fn()>()]. Of course that makes the rules in https://doc.rust-lang.org/std/option/ silly, but I would like to have some confirmation that NULL indeed goes to `None`.

Reply:
E.g. see https://c-faq.com/null/runtime0.html

Reply:
I couldn't figure this out in a reasonable amount of time so I've change the code into something conservative.